lib/report.js

```diff
@@ -158,7 +159,136 @@ class ConciseReporter extends Reporter {
   }
 }
 
+class TapReporter extends Reporter {
+  constructor(options) {
+    super(options);
+    this.successful = 0;
+    this.failed = 0;
+    this.skipped = 0;
+    this.testNumeration = [1];
+    this.lastIndex = 0;
+    this.message = '';
+    this.firstFailure = true;
+    this.got = [];
+    this.expected = [];
+  }
+
+  error(test, error) {
+    // remove one success because 'error' may only be received after end
+    // though the test may have already failed so we need a way to check
+    // if we should do this, i.e. store map of successful tests
+    // this.successes--;
+    // test.todo ? this.skipped++ : this.failures++;
+    // TODO(lundibundi): add test.metadata.filename here too once available
+    this.listFailure(
+      test,
+      { message: `Failure after end: ${error ? error.message : 'no message'}` },
+      ` Test failed: ${test.caption}`
+    );
+  }
+
+  lastElement(inc) {
+    if (inc) this.testNumeration[this.lastIndex]++;
+    return this.testNumeration[this.lastIndex];
+  }
+
+  calculateSpaces() {
+    const space = this.lastIndex * 2;
+    this.message = ' '.repeat(space);
+  }
+
+  parseTestResults(test) {
+    if (test) {
+      this.lastElement(1);
+      if (!test.success) {
+        this.listFailure(test, true);
+        this.got.forEach(el => {
+          this.message += `${' '.repeat(6)}- ${el}\n`;
+        });
+        this.message += `${' '.repeat(4)}expect: \n`;
+        this.expected.forEach(el => {
+          this.message += `${' '.repeat(6)}- ${el}\n`;
+        });
+        this.message += '  ...\n';
+        this.got = [];
+        this.expected = [];
+        console.log(this.message);
+      } else {
+        this.testNumeration.push(0);
+        this.lastIndex++;
+        this.calculateSpaces();
+        test.results.forEach(res => {
+          if (res.success) {
+            if (res.type === 'subtest') {
+              const t = res.test;
+              console.log(`${this.message}# Subtest: ${t.caption}`);
+              this.parseTestResults(t);
+              this.calculateSpaces();
+              const prefix = t.success ? 'ok' : 'not ok';
+              let message1 = `${this.message}${prefix} ${this.lastElement(0)}`;
+              message1 += ` - ${t.caption}`;
+              console.log(message1);
+            } else {
+              this.listSuccess(res);
+            }
+          }
+        });
+        console.log(`${this.message}1..${this.lastElement(0)}`);
+        this.lastIndex--;
+        this.testNumeration.pop();
+      }
+    }
+    if (!this.lastIndex) this.message = '\n';
+  }
+
+  listSuccess(test) {
+    const testPrefix = test.success ? 'ok' : 'not ok';
+    console.log(`${this.message}${testPrefix} ${this.lastElement(1)}`);
+  }
+
+  listFailure(test, firstFailure) {
+    test.results.forEach(res => {
+      if (firstFailure) {
+        firstFailure = false;
+        this.message += '  ---\n';
+        if (res.message) this.message += `  message: ${res.message}\n`;
```

Comment:
According to [TAP 13 specification](https://testanything.org/tap-version-13-specification.html) message should have either single or double quotation marks around it.
```suggestion
        if (res.message) this.message += `  message: '${res.message}'\n`;
```